src/Compilers/Core/Portable/CommandLine/EditorConfig.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents a single EditorConfig file, see http://editorconfig.org for details about the format.
+    /// </summary>
+    internal sealed class EditorConfig
+    {
+        // Matches EditorConfig section header such as "[*.{js,py}]", see http://editorconfig.org for details
+        private static readonly Regex _sectionMatcher = new Regex(@"^\s*\[(([^#;]|\\#|\\;)+)\]\s*([#;].*)?$", RegexOptions.Compiled);
+        // Matches EditorConfig property such as "indent_style = space", see http://editorconfig.org for details
+        private static readonly Regex _propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);
+
+        public Section GlobalSection { get; }
+
+        public string Directory { get; }
+
+        public ImmutableArray<Section> NamedSections { get; }
+
+        private EditorConfig(Section globalSection, ImmutableArray<Section> namedSections, string directory)
+        {
+            GlobalSection = globalSection;
+            NamedSections = namedSections;
+            Directory = directory;
+        }
+
+        /// <summary>
+        /// Gets whether this editorconfig is a the topmost editorconfig.
+        /// </summary>
+        public bool IsRoot
+            => GlobalSection.Properties.TryGetValue("root", out string val) && val == "true";
+
+        /// <summary>
+        /// Parses an editor config file text located within the given parent directory.
+        /// </summary>
+        public static EditorConfig Parse(string text, string parentDirectory)
```

Comment:
can this document the behavior that will happen on errors?  will this always succeed?  can there be parsing errors?

Reply:
oddly, this is called "parentdirectory" but is called "directory" everywhere else.  intentional?